ext/dom/node.c

```diff
@@ -1425,6 +1425,162 @@ PHP_METHOD(DOMNode, isSameNode)
+	{																											\
+		size_t counter = 0;																						\
+		while (node) {																							\
+			counter++;																							\
+			node = node->next;																					\
+		}																										\
+		return counter;																							\
+	}																											\
+	static bool PHP_DOM_FUNC_CAT(php_dom_node_list_equality_check, type)(const type *list1, const type *list2)	\
+	{																											\
+		size_t count = PHP_DOM_FUNC_CAT(php_dom_node_count_list_size, type)(list1);								\
+		if (count != PHP_DOM_FUNC_CAT(php_dom_node_count_list_size, type)(list2)) {								\
+			return false;																						\
+		}																										\
+		for (size_t i = 0; i < count; i++) {																	\
+			if (!php_dom_node_is_equal_node((const xmlNode *) list1, (const xmlNode *) list2)) {				\
+				return false;																					\
+			}																									\
+			list1 = list1->next;																				\
+			list2 = list2->next;																				\
+		}																										\
+		return true;																							\
+	}
+PHP_DOM_DEFINE_LIST_EQUALITY_HELPER(xmlNode)
+PHP_DOM_DEFINE_LIST_EQUALITY_HELPER(xmlNs)
+
+static bool php_dom_node_is_equal_node(const xmlNode *this, const xmlNode *other)
+{
+	ZEND_ASSERT(this != NULL);
+	ZEND_ASSERT(other != NULL);
+
+	if (this->type != other->type) {
+		return false;
+	}
+
+	/* Notes:
+	 *   - XML_DOCUMENT_TYPE_NODE is no longer created by libxml2, we only have to support XML_DTD_NODE.
+	 *   - element and attribute declarations are not exposed as nodes in DOM, so no comparison is needed for those. */
+	if (this->type == XML_ELEMENT_NODE) {
+		return xmlStrEqual(this->name, other->name)
+			&& php_dom_node_is_ns_prefix_equal(this, other)
+			&& php_dom_node_is_ns_uri_equal(this, other)
+			/* Check attributes first, then namespace declarations, then children */
+			&& php_dom_node_list_equality_check_xmlNode((const xmlNode *) this->properties, (const xmlNode *) other->properties)
+			&& php_dom_node_list_equality_check_xmlNs(this->nsDef, other->nsDef)
+			&& php_dom_node_list_equality_check_xmlNode(this->children, other->children);
+	} else if (this->type == XML_DTD_NODE) {
+		/* Note: in the living spec entity declarations and notations are no longer compared because they're considered obsolete. */
+		const xmlDtd *this_dtd = (const xmlDtd *) this;
+		const xmlDtd *other_dtd = (const xmlDtd *) other;
+		return xmlStrEqual(this_dtd->name, other_dtd->name)
+			&& xmlStrEqual(this_dtd->ExternalID, other_dtd->ExternalID)
+			&& xmlStrEqual(this_dtd->SystemID, other_dtd->SystemID);
+	} else if (this->type == XML_PI_NODE) {
+		return xmlStrEqual(this->name, other->name) && xmlStrEqual(this->content, other->content);
+	} else if (this->type == XML_TEXT_NODE || this->type == XML_COMMENT_NODE || this->type == XML_CDATA_SECTION_NODE) {
+		/* Note: spec doesn't explicitly mention it, but JS also checks equality of content for CDATA. */
```

Comment:
I don't know how the living spec works, but is it possible to raise a ticket there to explicitly mark this down? (I know Servo does this with the WebPlatformTests)

Reply:
Actually, https://dom.spec.whatwg.org/#cdatasection CDataSection implements the Text interface. So it is actually specified here: https://dom.spec.whatwg.org/#concept-node-equals under the Text section, I just misread it as literally meaning the Text node instead of the Text interface.
I'll remove the comment.